File: M12 NF1 (Jose Java)/Softlearning/model/Products/Product.php
<?php

include($_SERVER['DOCUMENT_ROOT'] . '/softlearning/Exceptions/BuildException.php');
include($_SERVER['DOCUMENT_ROOT'] . '/softlearning/model/operations/Check.php');
include_once 'Marketable.php';

abstract class Product implements Marketable {

    protected string $ident;
    protected float $price;
    protected bool $create = true;

    public function __construct(string $ident, float $price) {
        $errors = '';

        if ($this->setIdent($ident) !== 0) {
            $errors .= Check::getErrorMessage(Check::validate($ident)) . "\n";
        }

        if ($this->setPrice($price) !== 0) {
            $errors .= Check::getErrorMessage(Check::validate($price)) . "\n";
        }

        if (!empty($errors)) {
            throw new BuildException("Not possible to create the product: \n" . $errors);
        }
    }

    public function getIdent(): string {
        return $this->ident;
    }

    public function getPrice(): float {
        return $this->price;
    }

    public function hasDelayPay() {
        return $this->delayPay;
    }

    public function getDiscount() {
        return $this->discount;
    }

    public function getType() {
        return $this->type;
    }

    public function getPayMethod() {
        return $this->payMethod;
    }

    public function setIdent(string $ident): int {
        $errorIdent = Check::validate($ident);
        if ($errorIdent == 0) {
            $this->ident = $ident;
        }
        return $errorIdent;
    }

    public function setPrice(float $price): int {
        $errorPrice = Check::validate($price);
        if ($errorPrice == 0) {
            $this->price = $price;
        }
        return $errorPrice;
    }

    public abstract function getDetails(): mixed;
}




/*

abstract class Product implements Marketable {

    protected string $ident;
    protected float $price;
    protected bool $create=true;

    public function __construct(string $ident, float $price) {
        

        if ($this->setIdent($ident) !== 0) {
            $this->create = false;
        }

        
        if ($this->setPrice($price) !== 0) {
            $this->create = false;
        }


        if (!$this->create) {
            throw new BuildException("Not posible create the product(fail in product)");
        }
    }

    public function getIdent(): string {
        return $this->ident;
    }

    public function getPrice(): float {
        return $this->price;
    }

    public function hasDelayPay() {
        return $this->delayPay;
    }

    public function getDiscount() {
        return $this->discount;
    }

    public function getType() {
        return $this->getType;
    }

    public function getPayMethod() {
        return $this->payMethod;
    }

    public function setIdent(string $ident): int {
        $errorIdent = Check::validate($ident);
        if ($errorIdent != 0) {
            print Check::getErrorMessage($errorIdent);
        } else {
            $this->ident = $ident;
        }

        return $errorIdent;
    }

    public function setPrice(float $price): int {
        $errorPrice = Check::validate($price);
        if ($errorPrice != 0) {
            print Check::getErrorMessage($errorPrice);
        } else {
            $this->price = $price;
        }

        return $errorPrice;
    }

    public abstract function getDetails(): mixed;
}
*/

File: M12 Jose Java/softlearningmaven/src/main/java/com/core/entities/shared/products/Product.java
package com.core.entities.shared.products;

import com.core.checks.Check;
import com.core.entities.shared.marketable.Marketable;



public abstract class Product implements Marketable {  

    protected String ident;
    protected double price;
    

    // //estos son de marketable
    protected boolean delayPay;
    protected double discount;
    protected String type;
    protected String payMethod;

    protected Product(){//ESTO SERVIRA PARA MAS ADELANTE

    }
        //Las clases abstractas no pueden ser static, esta es la que se usa para chekear cuando se crea el libro
        public void product(String ident, double price, boolean delayPay, double discount, String type, String payMethod) throws Exception {
        StringBuilder errors = new StringBuilder();
        int errorCode;
        
        errorCode = setIdent(ident);
        if (errorCode != 0) {
            errors.append(Check.getErrorMessage(errorCode)).append("\n");
        }

        errorCode = setPrice(price);
        if (errorCode != 0) {
            errors.append(Check.getErrorMessage(errorCode)).append("\n");
        }

        errorCode = setDelayPay(delayPay);
        if (errorCode != 0 ) {
            errors.append(Check.getErrorMessage(errorCode)).append("\n");
        }

        errorCode = setDiscount(discount);
        if (errorCode != 0) {
            errors.append(Check.getErrorMessage(errorCode)).append("\n");
        }

        errorCode = setPayMethod(payMethod);
        if (errorCode != 0) {
            errors.append(Check.getErrorMessage(errorCode)).append("\n");
        }

        errorCode = setType(type);
        if (errorCode != 0) {
            errors.append(Check.getErrorMessage(errorCode)).append("\n");
        }
        
        
        if (errors.length() > 0) {
            throw new Exception("Not possible to create the product: \n" + errors.toString());
        }
    }

        
        
    

    public String getIdent() {
        return this.ident;
    }

    public double getPrice() {
        return this.price;
    }

    
    //son override porque son del Marketable
    @Override
    public boolean getDelayPay() {
        return delayPay;
    }

    @Override
    public double getDiscount() {
        return discount;
    }
    @Override
    public String getType() {
        return type;
    }
    @Override
    public String getPayMethod() {
        return payMethod;
    }
    
    //setters
    public int setDelayPay(boolean delayPay) {
        int errordelayPay = Check.checkBoolean(delayPay);
        if (errordelayPay == 0) {
            this.delayPay = delayPay;
        }
        return errordelayPay;
    }

    public int setDiscount(double discount) {
        int errorDiscount =Check.range(price);
        if(errorDiscount==0){
            this.discount = discount;

        }
        return errorDiscount;
    }

    public int setType(String type) {
        int errorType= Check.minMaxLength(type);
        if(errorType== 0){
            this.type = type;
        }
        return errorType;
    }

    public int setPayMethod(String payMethod) {
        int errorPayMethod = Check.minMaxLength(payMethod);
        if(errorPayMethod == 0){
            this.payMethod = payMethod;
        }
        return errorPayMethod;
    }

    public int setIdent(String ident) {
        int errorIdent = Check.minLenght(ident);
        if (errorIdent == 0) {
            this.ident = ident;
        }
        return errorIdent;
    }

    public int setPrice(double price) {
        int errorPrice = Check.range(price);
        if (errorPrice == 0) {
            this.price = price;
        }
        return errorPrice;
    }

    public abstract String getDetails();
}
